design/general_refactorings.md

```diff
@@ -0,0 +1,381 @@
+| :------------- | :----------------------------------------------------------- |
+| wrappers       | wrappers and utils of diffrent python libs :  arg_parse_wrapper, http_wrapper |
+| conjur_api     | files that talk directly with conjur. For Example   `conjur_client.py (conjur.py) , conjur_api.py (api.py), ssl_client.py (ssl_service)` |
+| logic          | all logic classes                                            |
+| controller     | all controller classes                                       |
+| logic          | all logic classes                                            |
+| data_objects   | all data objects classes                                     |
+
+
+
+### Improve Error Handling
+
+------
+
+##### Proposal
+
+This will help to better under the error handling flow. This will allow us to catch our own object and not general exceptions that do not proper helpful information to the end user. All Conjur CLI/SDK-specific exceptions should be documented in the funtions that call them.
+
+##### Current state
+
+Through the codebase, we have general errors that are not representative of the actual error that took place. The exception is not documented in the fucntion documentation in the code.
+For example, in the following `get_certificate` function, we raise a general exception if anything goes wrong during the certificate retrieval operation.
+
+```python
+def get_certificate(self, hostname, port):
+    """
+    Method for connecting to Conjur to fetch the certificate chain
+    """
+    if port is None:
+        port = DEFAULT_PORT
+    try:
+        fingerprint, readable_certificate = self.ssl_service.get_certificate(hostname, port)
+        logging.debug("Successfully fetched certificate")
+    except Exception as error:
+        raise Exception(f"Unable to retrieve certificate from {hostname}:{port}. " \
+                        f"Reason: {str(error)}") from error
+
+    return fingerprint, readable_certificate
+```
+
+We raise Exception if anything happens in certificate retrivial.
+
+##### Recommendation
+
+Map current error and exception handling and create Conjur-specific exceptions. The code block above should be:
+
+```python
+    def get_certificate(self, hostname, port):
+        """
+        Method for connecting to Conjur to fetch the certificate chain
+        :raises CertificateRetrivialFailed when certificate retrivial failes
+        """
+        if port is None:
+            port = DEFAULT_PORT
+        try:
+            fingerprint, readable_certificate = self.ssl_service.get_certificate(hostname, port)
+            logging.debug("Successfully fetched certificate")
+        except Exception as error:
+            raise CertificateRetrievalFailedException(f"Unable to retrieve certificate from {hostname}:{port}. " \
+                            f"Reason: {str(error)}") from error
+
+        return fingerprint, readable_certificate
+```
+
+Appendinx 1 lists the error handling cases where fixes need to be done.
```

Comment:
Appendix

Reply:
Appendix 1: List of error handling cases

Reply:
"See Appendix 1 for a list of error handling where fixes are required"

Please provide link to the appendix 1. See [here](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet) for how

Reply:
Added